File: data/src/main/java/com/arturogutierrez/openticator/storage/CategoryDiskDataStore.java
package com.arturogutierrez.openticator.storage;

import com.arturogutierrez.openticator.domain.account.model.Account;
import com.arturogutierrez.openticator.domain.category.model.Category;
import com.arturogutierrez.openticator.domain.category.repository.CategoryDataStore;
import com.arturogutierrez.openticator.storage.realm.mapper.CategoryRealmMapper;
import com.arturogutierrez.openticator.storage.realm.model.AccountRealm;
import com.arturogutierrez.openticator.storage.realm.model.CategoryRealm;
import io.realm.Realm;
import io.realm.RealmResults;
import java.util.List;
import javax.inject.Inject;
import rx.Observable;
import rx.subjects.PublishSubject;
import rx.subjects.Subject;

public class CategoryDiskDataStore implements CategoryDataStore {

  private final CategoryRealmMapper categoryRealmMapper;
  private final Subject<Void, Void> changesPublishSubject;

  @Inject
  public CategoryDiskDataStore(CategoryRealmMapper categoryRealmMapper) {
    this.categoryRealmMapper = categoryRealmMapper;
    this.changesPublishSubject = PublishSubject.<Void>create().toSerialized();
  }

  @Override
  public Observable<Category> add(Category category) {
    Observable<Category> categoryObservable = Observable.fromCallable(() -> {
      CategoryRealm categoryRealm = categoryRealmMapper.transform(category);

      Realm defaultRealm = Realm.getDefaultInstance();
      defaultRealm.executeTransaction(realm -> realm.copyToRealm(categoryRealm));

      return category;
    });

    return categoryObservable.doOnNext(categoryAdded -> notifyAccountChanges());
  }

  @Override
  public Observable<Category> addAccount(Category category, Account account) {
    return Observable.fromCallable(() -> {
      Realm defaultRealm = Realm.getDefaultInstance();
      defaultRealm.executeTransaction(realm -> {
        CategoryRealm categoryRealm = getCategoryAsBlocking(realm, category.getCategoryId());
        AccountRealm accountRealm = getAccountAsBlocking(realm, account.getAccountId());
        if (categoryRealm == null || accountRealm == null) {
          return;
        }

        accountRealm.setCategory(categoryRealm);
      });

      return category;
    });
  }

  @Override
  public Observable<List<Category>> getCategories() {
    return changesPublishSubject.map(aVoid -> getCategoriesAsBlocking())
        .startWith(Observable.just(getCategoriesAsBlocking()));
  }

  private List<Category> getCategoriesAsBlocking() {
    Realm realm = Realm.getDefaultInstance();
    realm.refresh();
    RealmResults<CategoryRealm> realmResults =
        realm.where(CategoryRealm.class).findAllSorted("name");
    return categoryRealmMapper.reverseTransform(realmResults);
  }

  private CategoryRealm getCategoryAsBlocking(Realm realm, String categoryId) {
    return realm.where(CategoryRealm.class).equalTo("categoryId", categoryId).findFirst();
  }

  private AccountRealm getAccountAsBlocking(Realm realm, String accountId) {
    return realm.where(AccountRealm.class).equalTo("accountId", accountId).findFirst();
  }

  private void notifyAccountChanges() {
    changesPublishSubject.onNext(null);
  }
}

File: data/src/main/kotlin/com/arturogutierrez/openticator/storage/CategoryDiskDataStore.kt
package com.arturogutierrez.openticator.storage

import com.arturogutierrez.openticator.domain.account.model.Account
import com.arturogutierrez.openticator.domain.category.model.Category
import com.arturogutierrez.openticator.domain.category.repository.CategoryDataStore
import com.arturogutierrez.openticator.storage.realm.mapper.CategoryRealmMapper
import com.arturogutierrez.openticator.storage.realm.model.AccountRealm
import com.arturogutierrez.openticator.storage.realm.model.CategoryRealm
import io.realm.Realm
import rx.Observable
import rx.subjects.PublishSubject
import javax.inject.Inject

class CategoryDiskDataStore @Inject constructor(private val categoryRealmMapper: CategoryRealmMapper) : CategoryDataStore {

  private val changesPublishSubject = PublishSubject.create<Void>().toSerialized()

  override fun add(category: Category): Observable<Category> {
    val categoryObservable = Observable.fromCallable {
      val categoryRealm = categoryRealmMapper.transform(category)

      val defaultRealm = Realm.getDefaultInstance()
      defaultRealm.executeTransaction { realm -> realm.copyToRealm(categoryRealm) }
      defaultRealm.close()

      category
    }

    return categoryObservable.doOnNext { categoryAdded -> notifyAccountChanges() }
  }

  override fun addAccount(category: Category, account: Account): Observable<Category> {
    return Observable.fromCallable {
      val defaultRealm = Realm.getDefaultInstance()
      defaultRealm.executeTransaction { realm ->
        val categoryRealm = getCategoryAsBlocking(realm, category.categoryId)
        val accountRealm = getAccountAsBlocking(realm, account.accountId)
        if (categoryRealm == null || accountRealm == null) {
          return@executeTransaction
        }

        accountRealm.category = categoryRealm
      }
      defaultRealm.close()

      category
    }
  }

  override val categories: Observable<List<Category>>
    get() = changesPublishSubject.map { categoriesAsBlocking }
        .startWith(Observable.just(categoriesAsBlocking))

  private val categoriesAsBlocking: List<Category>
    get() {
      val realm = Realm.getDefaultInstance()
      val realmResults = realm.where(CategoryRealm::class.java).findAllSorted("name")
      val categories = categoryRealmMapper.reverseTransform(realmResults)
      realm.close()

      return categories
    }

  private fun getCategoryAsBlocking(realm: Realm, categoryId: String): CategoryRealm? {
    return realm.where(CategoryRealm::class.java).equalTo("categoryId", categoryId).findFirst()
  }

  private fun getAccountAsBlocking(realm: Realm, accountId: String): AccountRealm? {
    return realm.where(AccountRealm::class.java).equalTo("accountId", accountId).findFirst()
  }

  private fun notifyAccountChanges() {
    changesPublishSubject.onNext(null)
  }
}
